File: packages/coding-agent/src/modes/interactive/components/sidebar.ts
import { type Component, truncateToWidth, visibleWidth } from "@earendil-works/pi-tui";
import type { ContextUsage } from "../../../core/extensions/types.ts";
import type { RuntimeStateSnapshot } from "../../../core/runtime-state.ts";
import { theme } from "../theme/theme.ts";
import type { SidebarPermissionComponent } from "./sidebar-permission.ts";

function fmtNumber(value: number | undefined): string {
	if (!value) return "0";
	if (value < 1000) return String(value);
	if (value < 1000000) return `${(value / 1000).toFixed(value < 10000 ? 1 : 0)}k`;
	return `${(value / 1000000).toFixed(1)}M`;
}

function fmtDuration(ms: number | undefined): string {
	if (!ms) return "0s";
	if (ms < 60000) return `${Math.round(ms / 1000)}s`;
	return `${Math.round(ms / 60000)}m`;
}

export class SidebarComponent implements Component {
	private getState: () => RuntimeStateSnapshot;
	private getContextUsage: () => ContextUsage | undefined;
	private permissionComponent: SidebarPermissionComponent | null = null;

	constructor(getState: () => RuntimeStateSnapshot, getContextUsage: () => ContextUsage | undefined) {
		this.getState = getState;
		this.getContextUsage = getContextUsage;
	}

	setPermissionComponent(comp: SidebarPermissionComponent | null): void {
		this.permissionComponent = comp;
	}

	invalidate(): void {}

	render(width: number): string[] {
		if (width < 16) {
			return [];
		}
		const panel = this.renderPanel(width);
		// 在底部追加权限审批（如果有）
		if (this.permissionComponent?.isActive()) {
			const permLines = this.permissionComponent.render(width);
			panel.push(...permLines);
		}
		return panel;
	}

	private renderPanel(width: number): string[] {
		const state = this.getState();
		const ctx = this.getContextUsage();
		const b = theme.fg("borderMuted", "│");
		const inner = width - 2;
		const lines: string[] = [];

		// ─── PROJECT TITLE ───────────────────
		const projectName = (state.project.sessionName ?? state.project.projectName).toUpperCase();
		lines.push(`${b} ${truncateToWidth(theme.bold(theme.fg("accent", projectName)), inner, "...")}`);

		// git
		if (state.gitStatus.enabled) {
			const branch = state.gitStatus.branch ?? "detached";
			const changes = state.gitStatus.uncommittedChangesCount;
			const changesStr = changes > 0 ? ` ·${theme.fg("success", String(changes))} changes` : "";
			lines.push(`${b} ${truncateToWidth(branch + changesStr, inner, "...")}`);
		}

		// model
		const modelName = this.formatModelName(state.modelUsage.model ?? "no-model");
		const thinkLevel = state.modelUsage.reasoningEffort ?? "off";
		const thinkStr = thinkLevel.charAt(0).toUpperCase() + thinkLevel.slice(1);
		lines.push(`${b} ${truncateToWidth(`${modelName}|${thinkStr}`, inner, "...")}`);

		// mode
		const permFull =
			state.permission.mode === "free-permission"
				? theme.fg("warning", "free")
				: state.permission.mode === "auto-permission"
					? theme.fg("success", "auto")
					: "on-request";
		const editModeStr =
			state.editMode === "think_first" ? theme.fg("warning", "think_first") : theme.fg("dim", "normal");
		lines.push(`${b} ▶ ${permFull} · ${editModeStr}`);
		lines.push(b);

		// ─── CONTEXT ─────────────────────────
		const ctxPercent = ctx?.percent ?? 0;
		const ctxLabel = ctx?.percent !== null ? `${Math.round(ctxPercent)}%` : "—";
		lines.push(`${b} ${theme.fg("muted", "CONTEXT")}`);
		lines.push(this.progressBar(width, ctxPercent, ctxLabel));
		lines.push(b);

		// ─── TOKENS ──────────────────────────
		const tokIn = fmtNumber(state.modelUsage.promptTokens);
		const tokOut = fmtNumber(state.modelUsage.completionTokens);
		const tokTotal = fmtNumber(state.modelUsage.sessionTotalTokens);
		lines.push(`${b} ${theme.fg("muted", "TOKENS")} ${theme.fg("dim", `[${tokTotal}]`)}`);
		lines.push(`${b} In ${tokIn} · Out ${tokOut}`);
		lines.push(b);

		// ─── AGENTS（有数据时显示，高亮当前查看的 agent）─────────
		if (state.activeSubagents.length > 0) {
			lines.push(`${b} ${theme.fg("muted", "AGENTS")}`);
			const viewingMain = !state.viewingSubagentId;
			const mainLabel = viewingMain ? theme.bold(theme.fg("accent", "main")) : "main";
			const mainIcon = viewingMain ? theme.fg("accent", "▶") : theme.fg("success", "●");
			lines.push(`${b} ${mainIcon} ${mainLabel}`);
			for (const agent of state.activeSubagents.slice(0, 5)) {
				const elapsed = fmtDuration(agent.elapsedTime ?? (agent.startTime ? Date.now() - agent.startTime : 0));
				const icon =
					agent.id === state.viewingSubagentId
						? theme.fg("accent", "▶")
						: this.agentIcon(agent.status, agent.errorMessage);
				const name = agent.id === state.viewingSubagentId ? theme.bold(theme.fg("accent", agent.name)) : agent.name;
				const suffix = agent.status === "completed" ? theme.fg("dim", "(done)") : theme.fg("dim", elapsed);
				lines.push(`${b} ${icon} ${truncateToWidth(name, inner - 12, "...")} ${suffix}`);
			}
			if (state.activeSubagents.length > 5) {
				lines.push(`${b} ${theme.fg("dim", `  …${state.activeSubagents.length - 5} more`)}`);
			}
			lines.push(b);
		}

		// ─── FILES（项目级未提交变更）────────────────────
		if (state.gitStatus.enabled && state.gitStatus.uncommittedFiles?.length) {
			const totalAdd = state.gitStatus.uncommittedFiles.reduce((s, f) => s + (f.additions ?? 0), 0);
			const totalDel = state.gitStatus.uncommittedFiles.reduce((s, f) => s + (f.deletions ?? 0), 0);
			const statStr = [
				totalAdd > 0 ? theme.fg("success", `+${totalAdd}`) : "",
				totalDel > 0 ? theme.fg("error", `-${totalDel}`) : "",
			]
				.filter(Boolean)
				.join(" ");
			const header = statStr
				? `${theme.fg("muted", "PROJECT FILES")} ${statStr}`
				: theme.fg("muted", "PROJECT FILES");
			lines.push(`${b} ${header}`);
			for (const file of state.gitStatus.uncommittedFiles.slice(0, 10)) {
				const name = file.path.split("/").pop() ?? file.path;
				const icon =
					file.status === "added"
						? theme.fg("success", "+")
						: file.status === "deleted"
							? theme.fg("error", "−")
							: theme.fg("warning", "~");
				const lineStats =
					file.additions !== undefined || file.deletions !== undefined
						? " " +
							(file.additions !== undefined ? theme.fg("success", `+${file.additions}`) : "") +
							(file.deletions !== undefined ? theme.fg("error", `-${file.deletions}`) : "")
						: "";
				lines.push(`${b} ${icon} ${truncateToWidth(name, inner - 3, "...")}${lineStats}`);
			}
			if (state.gitStatus.uncommittedFiles.length > 10) {
				lines.push(`${b} ${theme.fg("dim", `  …${state.gitStatus.uncommittedFiles.length - 10} more`)}`);
			}
			lines.push(b);
		}

		// ─── FILES（当前会话变更）──────────────────────
		if (state.changedFiles.length > 0) {
			const totalAdd = state.changedFiles.reduce((s, f) => s + (f.additions ?? 0), 0);
			const totalDel = state.changedFiles.reduce((s, f) => s + (f.deletions ?? 0), 0);
			const statStr = [
				totalAdd > 0 ? theme.fg("success", `+${totalAdd}`) : "",
				totalDel > 0 ? theme.fg("error", `-${totalDel}`) : "",
			]
				.filter(Boolean)
				.join(" ");
			const filesHeader = statStr
				? `${theme.fg("muted", "SESSION FILES")} ${statStr}`
				: theme.fg("muted", "SESSION FILES");
			lines.push(`${b} ${filesHeader}`);
			for (const file of state.changedFiles.slice(0, 5)) {
				const name = file.path.split("/").pop() ?? file.path;
				const icon =
					file.status === "added"
						? theme.fg("success", "+")
						: file.status === "deleted"
							? theme.fg("error", "−")
							: theme.fg("warning", "~");
				lines.push(`${b} ${icon} ${truncateToWidth(name, inner - 3, "...")}`);
			}
			if (state.changedFiles.length > 5) {
				lines.push(`${b} ${theme.fg("dim", `  …${state.changedFiles.length - 5} more`)}`);
			}
			lines.push(b);
		}

		// ─── TOOLS（有调用时才显示）──────────
		if (state.toolCallStats.length > 0) {
			lines.push(`${b} ${theme.fg("muted", "TOOLS")}`);
			const sorted = [...state.toolCallStats].sort((a, b) => b.callCount - a.callCount).slice(0, 4);
			const toolParts = sorted.map((t) => `${t.toolName} ×${t.callCount}`);
			let current = "";
			for (const part of toolParts) {
				if (current.length === 0) {
					current = part;
				} else if (current.length + 2 + part.length <= inner - 1) {
					current += `  ${part}`;
				} else {
					lines.push(`${b} ${current}`);
					current = part;
				}
			}
			if (current.length > 0) {
				lines.push(`${b} ${current}`);
			}
			lines.push(b);
		}

		return lines.map((line) => {
			const vis = visibleWidth(line);
			if (vis < width) return line + " ".repeat(width - vis);
			return truncateToWidth(line, width, "...");
		});
	}

	private progressBar(width: number, percent: number, label: string): string {
		const b = theme.fg("borderMuted", "│");
		const labelWidth = label.length + 1;
		const barWidth = width - 3 - labelWidth;
		const filled = Math.round((percent / 100) * barWidth);
		const empty = barWidth - filled;

		let color: "success" | "warning" | "error" = "success";
		if (percent > 90) color = "error";
		else if (percent > 70) color = "warning";

		const labelStr = this.colorByPercent(label, percent);
		return `${b} ${theme.fg(color, "▓".repeat(filled))}${theme.fg("dim", "░".repeat(empty))} ${labelStr}`;
	}

	private colorByPercent(text: string, percent: number): string {
		if (percent > 90) return theme.fg("error", text);
		if (percent > 70) return theme.fg("warning", text);
		return theme.fg("success", text);
	}

	private agentIcon(status: string, errorMessage?: string): string {
		switch (status) {
			case "completed":
				return theme.fg("success", "●");
			case "failed":
				return theme.fg("error", "○");
			case "running":
				if (errorMessage?.includes("permission")) {
					return theme.fg("warning", "○");
				}
				return "○";
			default:
				return "●";
		}
	}

	private formatModelName(model: string): string {
		const parts = model.split(".");
		let name = parts[parts.length - 1];
		name = name.replace(/-v\d+$/, "");
		name = name.replace(/(\d+)-(\d+)$/, "$1.$2");
		return name;
	}
}

File: legacy/rozsa-tui/src/panels/sidebar.rs
// components/sidebar.rs — 侧边栏渲染
//
// Internal Framework:
// sidebar.rs
// ├── render_sidebar()  侧边栏渲染 (git/model/tokens/agents/files)
// └── truncate()        文本截断工具
//
// Related Docs:
// - [TUI Design](../../../docs/rozsa_framework.md#rozsa-tui--ratatui-终端前端)

// components/sidebar.rs — 侧边栏渲染：项目信息、模型状态、上下文使用率、代理队列、工具统计等
//
// Internal Framework:
// sidebar.rs
// ├── render_sidebar()     pub fn 主渲染入口
// └── truncate()           pub fn 文本截断（考虑 unicode width）
//
// Related Docs:
// - [TUI Design](../../../docs/rozsa_framework.md#rozsa-tui--ratatui-终端前端)

use std::cmp::Reverse;

use ratatui::{
    layout::Rect,
    style::{Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Wrap},
};
use serde_json::Value;

use crate::{app::AppState, backend::SubagentView, theme::THEME};

pub fn render_sidebar(
    frame: &mut ratatui::Frame<'_>,
    area: Rect,
    state: &AppState,
    agents: Option<&dyn SubagentView>,
) {
    let ui = &state.ui;
    let runtime = ui.runtime_state.as_ref();
    let inner = area.width.saturating_sub(2) as usize;
    let mut lines = Vec::<Line<'static>>::new();

    let project_name = runtime
        .and_then(|state| state.pointer("/project/sessionName"))
        .and_then(Value::as_str)
        .or_else(|| {
            runtime
                .and_then(|state| state.pointer("/project/projectName"))
                .and_then(Value::as_str)
        })
        .or(ui.session_name.as_deref())
        .unwrap_or(ui.app_name.as_str())
        .to_uppercase();
    lines.push(Line::styled(
        truncate(&project_name, inner),
        Style::default()
            .fg(THEME.accent)
            .add_modifier(Modifier::BOLD),
    ));

    append_git(&mut lines, runtime, inner);
    append_model(&mut lines, runtime, ui, inner);
    append_mode(&mut lines, runtime);
    lines.push(Line::raw(""));
    append_context(&mut lines, ui, inner);
    lines.push(Line::raw(""));
    append_tokens(&mut lines, runtime);
    append_queue(&mut lines, ui, inner);
    append_agents(&mut lines, agents, inner);
    append_files(&mut lines, runtime, inner);
    append_tools(&mut lines, runtime, inner);
    append_notices(&mut lines, state, inner);

    let paragraph = Paragraph::new(lines).wrap(Wrap { trim: false }).block(
        Block::default()
            .borders(Borders::LEFT)
            .border_style(Style::default().fg(THEME.border_muted)),
    );
    frame.render_widget(paragraph, area);
}

fn append_git(lines: &mut Vec<Line<'static>>, runtime: Option<&Value>, inner: usize) {
    if !runtime
        .and_then(|state| state.pointer("/gitStatus/enabled"))
        .and_then(Value::as_bool)
        .unwrap_or(false)
    {
        return;
    }
    let branch = runtime
        .and_then(|state| state.pointer("/gitStatus/branch"))
        .and_then(Value::as_str)
        .unwrap_or("detached");
    let changes = runtime
        .and_then(|state| state.pointer("/gitStatus/uncommittedChangesCount"))
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let mut spans = vec![Span::raw(truncate(branch, inner.saturating_sub(10)))];
    if changes > 0 {
        spans.push(Span::raw(" ·"));
        spans.push(Span::styled(
            changes.to_string(),
            Style::default().fg(THEME.assistant_msg),
        ));
        spans.push(Span::raw(" changes"));
    }
    lines.push(Line::from(spans));
}

fn append_model(
    lines: &mut Vec<Line<'static>>,
    runtime: Option<&Value>,
    ui: &crate::protocol::NativeUiState,
    inner: usize,
) {
    let model_name = runtime
        .and_then(|state| state.pointer("/modelUsage/model"))
        .and_then(Value::as_str)
        .or(ui.model.as_ref().map(|model| model.id.as_str()))
        .unwrap_or("no-model");
    let thinking = runtime
        .and_then(|state| state.pointer("/modelUsage/reasoningEffort"))
        .and_then(Value::as_str)
        .unwrap_or(&ui.thinking_level);
    lines.push(Line::raw(truncate(
        &format!("{}|{}", format_model_name(model_name), title_case(thinking)),
        inner,
    )));
}

fn append_mode(lines: &mut Vec<Line<'static>>, runtime: Option<&Value>) {
    let permission = runtime
        .and_then(|state| state.pointer("/permission/mode"))
        .and_then(Value::as_str)
        .unwrap_or("on-request");
    let permission_label = match permission {
        "free-permission" => "free",
        "auto-permission" => "auto",
        _ => "on-request",
    };
    let edit_mode = runtime
        .and_then(|state| state.pointer("/editMode"))
        .and_then(Value::as_str)
        .unwrap_or("normal");
    lines.push(Line::from(vec![
        Span::styled("▶ ", Style::default().fg(THEME.accent)),
        Span::styled(
            permission_label,
            Style::default().fg(if permission_label == "auto" {
                THEME.success
            } else {
                THEME.warning
            }),
        ),
        Span::raw(" · "),
        Span::styled(edit_mode.to_string(), Style::default().fg(THEME.muted)),
    ]));
}

fn append_context(
    lines: &mut Vec<Line<'static>>,
    ui: &crate::protocol::NativeUiState,
    inner: usize,
) {
    let percent = ui
        .context_usage
        .as_ref()
        .and_then(|ctx| ctx.get("percent"))
        .and_then(Value::as_f64);
    let label = percent
        .map(|value| format!("{value:.0}%"))
        .unwrap_or_else(|| "—".to_string());
    lines.push(Line::styled("CONTEXT", Style::default().fg(THEME.muted)));
    lines.push(progress_line(percent.unwrap_or(0.0), &label, inner));
    lines.push(Line::raw(""));
}

fn append_tokens(lines: &mut Vec<Line<'static>>, runtime: Option<&Value>) {
    let prompt = runtime
        .and_then(|state| state.pointer("/modelUsage/promptTokens"))
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let completion = runtime
        .and_then(|state| state.pointer("/modelUsage/completionTokens"))
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let session = runtime
        .and_then(|state| state.pointer("/modelUsage/sessionTotalTokens"))
        .and_then(Value::as_u64)
        .unwrap_or(0);
    lines.push(Line::from(vec![
        Span::styled("TOKENS", Style::default().fg(THEME.muted)),
        Span::styled(
            format!(" [{}]", fmt_number(session)),
            Style::default().fg(THEME.muted),
        ),
    ]));
    lines.push(Line::raw(format!(
        "In {} · Out {}",
        fmt_number(prompt),
        fmt_number(completion)
    )));
}

fn append_queue(lines: &mut Vec<Line<'static>>, ui: &crate::protocol::NativeUiState, inner: usize) {
    if ui.pending_messages.is_empty() {
        return;
    }
    lines.push(Line::raw(""));
    lines.push(Line::styled("QUEUE", Style::default().fg(THEME.muted)));
    for message in ui.pending_messages.iter().take(5) {
        lines.push(Line::raw(format!(
            "- {}",
            truncate(message, inner.saturating_sub(2))
        )));
    }
}

fn append_agents(lines: &mut Vec<Line<'static>>, agents: Option<&dyn SubagentView>, inner: usize) {
    let Some(view) = agents else {
        return;
    };
    let subagents = view.list_subagents_sync();
    if subagents.is_empty() {
        return;
    }
    let viewing = view.viewing_subagent_id_sync();
    lines.push(Line::raw(""));
    lines.push(Line::styled("AGENTS", Style::default().fg(THEME.muted)));
    lines.push(Line::styled(
        if viewing.is_none() {
            "▶ main"
        } else {
            "● main"
        },
        Style::default().fg(THEME.accent),
    ));
    for agent in subagents.iter().take(5) {
        let icon = if Some(&agent.id) == viewing.as_ref() {
            "▶"
        } else {
            match agent.status {
                rozsa_app::subagent::SubagentStatus::Running => "○",
                _ => "●",
            }
        };
        let status_str = format!("{:?}", agent.status).to_lowercase();
        let name_max = inner.saturating_sub(status_str.len() + 4);
        lines.push(Line::raw(format!(
            "{icon} {} {}",
            truncate(&agent.name, name_max),
            status_str
        )));
    }
    if subagents.len() > 5 {
        lines.push(Line::styled(
            format!("  …{} more", subagents.len() - 5),
            Style::default().fg(THEME.muted),
        ));
    }
}

fn append_files(lines: &mut Vec<Line<'static>>, runtime: Option<&Value>, inner: usize) {
    if let Some(files) = runtime
        .and_then(|state| state.pointer("/gitStatus/uncommittedFiles"))
        .and_then(Value::as_array)
    {
        if !files.is_empty() {
            append_file_section(lines, "PROJECT FILES", files, inner, 10);
        }
    }
    if let Some(files) = runtime
        .and_then(|state| state.pointer("/changedFiles"))
        .and_then(Value::as_array)
    {
        if !files.is_empty() {
            append_file_section(lines, "SESSION FILES", files, inner, 5);
        }
    }
}

fn append_tools(lines: &mut Vec<Line<'static>>, runtime: Option<&Value>, inner: usize) {
    let Some(tools) = runtime
        .and_then(|state| state.pointer("/toolCallStats"))
        .and_then(Value::as_array)
    else {
        return;
    };
    if tools.is_empty() {
        return;
    }
    let mut sorted = tools.iter().collect::<Vec<_>>();
    sorted.sort_by_key(|tool| Reverse(tool.get("callCount").and_then(Value::as_u64).unwrap_or(0)));
    lines.push(Line::raw(""));
    lines.push(Line::styled("TOOLS", Style::default().fg(THEME.muted)));
    let mut current = String::new();
    for tool in sorted.iter().take(4) {
        let name = tool
            .get("toolName")
            .and_then(Value::as_str)
            .unwrap_or("tool");
        let count = tool.get("callCount").and_then(Value::as_u64).unwrap_or(0);
        let part = format!("{name} ×{count}");
        if current.is_empty() {
            current = part;
        } else if current.len() + part.len() + 2 <= inner {
            current.push_str("  ");
            current.push_str(&part);
        } else {
            lines.push(Line::raw(truncate(&current, inner)));
            current = part;
        }
    }
    if !current.is_empty() {
        lines.push(Line::raw(truncate(&current, inner)));
    }
}

fn append_notices(lines: &mut Vec<Line<'static>>, state: &AppState, inner: usize) {
    if state.notifications.is_empty() {
        return;
    }
    lines.push(Line::raw(""));
    lines.push(Line::styled("NOTICES", Style::default().fg(THEME.muted)));
    for notice in state.notifications.iter().rev().take(3) {
        let color = match notice.level.as_str() {
            "error" => THEME.error,
            "warning" => THEME.warning,
            _ => THEME.muted,
        };
        lines.push(Line::styled(
            truncate(&notice.message, inner),
            Style::default().fg(color),
        ));
    }
}

fn append_file_section(
    lines: &mut Vec<Line<'static>>,
    title: &str,
    files: &[Value],
    inner: usize,
    limit: usize,
) {
    let total_add = files
        .iter()
        .filter_map(|file| file.get("additions").and_then(Value::as_u64))
        .sum::<u64>();
    let total_del = files
        .iter()
        .filter_map(|file| file.get("deletions").and_then(Value::as_u64))
        .sum::<u64>();
    lines.push(Line::raw(""));
    lines.push(Line::from(vec![
        Span::styled(title.to_string(), Style::default().fg(THEME.muted)),
        if total_add > 0 {
            Span::styled(
                format!(" +{total_add}"),
                Style::default().fg(THEME.assistant_msg),
            )
        } else {
            Span::raw("")
        },
        if total_del > 0 {
            Span::styled(format!(" -{total_del}"), Style::default().fg(THEME.error))
        } else {
            Span::raw("")
        },
    ]));
    for file in files.iter().take(limit) {
        let path = file.get("path").and_then(Value::as_str).unwrap_or("file");
        let name = path.rsplit('/').next().unwrap_or(path);
        let icon = match file
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or("modified")
        {
            "added" => "+",
            "deleted" => "-",
            _ => "~",
        };
        let add = file.get("additions").and_then(Value::as_u64).unwrap_or(0);
        let del = file.get("deletions").and_then(Value::as_u64).unwrap_or(0);
        let stats_str = match (add, del) {
            (0, 0) => String::new(),
            (a, 0) => format!(" +{a}"),
            (0, d) => format!(" -{d}"),
            (a, d) => format!(" +{a}-{d}"),
        };
        let name_max = inner.saturating_sub(2 + stats_str.len());
        let mut spans = vec![
            Span::styled(format!("{icon} "), Style::default().fg(THEME.muted)),
            Span::raw(truncate(name, name_max)),
        ];
        if add > 0 {
            spans.push(Span::styled(
                format!(" +{add}"),
                Style::default().fg(THEME.assistant_msg),
            ));
        }
        if del > 0 {
            spans.push(Span::styled(
                format!("-{del}"),
                Style::default().fg(THEME.error),
            ));
        }
        lines.push(Line::from(spans));
    }
    if files.len() > limit {
        lines.push(Line::styled(
            format!("  …{} more", files.len() - limit),
            Style::default().fg(THEME.muted),
        ));
    }
}

fn progress_line(percent: f64, label: &str, inner: usize) -> Line<'static> {
    let safe_percent = percent.clamp(0.0, 100.0);
    let label_width = label.chars().count() + 1;
    let bar_width = inner.saturating_sub(label_width).max(1);
    let filled = ((safe_percent / 100.0) * bar_width as f64).round() as usize;
    let empty = bar_width.saturating_sub(filled);
    let color = if safe_percent > 90.0 {
        THEME.error
    } else if safe_percent > 70.0 {
        THEME.warning
    } else {
        THEME.assistant_msg
    };
    Line::from(vec![
        Span::styled("▓".repeat(filled), Style::default().fg(color)),
        Span::styled("░".repeat(empty), Style::default().fg(THEME.muted)),
        Span::raw(" "),
        Span::styled(label.to_string(), Style::default().fg(color)),
    ])
}

fn fmt_number(value: u64) -> String {
    if value < 1_000 {
        value.to_string()
    } else if value < 1_000_000 {
        let value = value as f64 / 1_000.0;
        if value < 10.0 {
            format!("{value:.1}k")
        } else {
            format!("{value:.0}k")
        }
    } else {
        format!("{:.1}M", value as f64 / 1_000_000.0)
    }
}

fn title_case(text: &str) -> String {
    let mut chars = text.chars();
    let Some(first) = chars.next() else {
        return String::new();
    };
    format!("{}{}", first.to_uppercase(), chars.as_str())
}

fn format_model_name(model: &str) -> String {
    let mut name = model.rsplit('.').next().unwrap_or(model).to_string();
    if let Some(stripped) = name.strip_suffix("-v1") {
        name = stripped.to_string();
    }
    name
}

pub fn truncate(text: &str, max_width: usize) -> String {
    use unicode_width::UnicodeWidthChar;
    let mut width = 0;
    let mut out = String::new();
    for ch in text.chars() {
        let w = ch.width().unwrap_or(0);
        if width + w > max_width {
            out.push('…');
            return out;
        }
        out.push(ch);
        width += w;
    }
    out
}
